yt_dlp/extractor/qqmusic.py

```diff
@@ -1,245 +1,328 @@
+import base64
+import functools
+import json
 import random
-import re
 import time
 
 from .common import InfoExtractor
 from ..utils import (
     ExtractorError,
+    OnDemandPagedList,
     clean_html,
+    int_or_none,
+    join_nonempty,
+    js_to_json,
+    str_or_none,
     strip_jsonp,
+    traverse_obj,
     unescapeHTML,
+    url_or_none,
+    urljoin,
 )
 
 
-class QQMusicIE(InfoExtractor):
+class QQMusicBaseIE(InfoExtractor):
+    def _get_cookie(self, key, default=None):
+        return getattr(self._get_cookies('https://y.qq.com').get(key), 'value', default)
+
+    def _get_g_tk(self):
+        n = 5381
+        for c in self._get_cookie('qqmusic_key', ''):
+            n += (n << 5) + ord(c)
+        return n & 2147483647
+
+    def _get_uin(self):
+        return int_or_none(self._get_cookie('uin')) or 0
+
+    @property
+    def is_logged_in(self):
+        return bool(self._get_uin() and self._get_cookie('fqm_pvqid'))
+
+    # Reference: m_r_GetRUin() in top_player.js
+    # http://imgcache.gtimg.cn/music/portal_v3/y/top_player.js
+    @staticmethod
+    def m_r_get_ruin():
+        cur_ms = int(time.time() * 1000) % 1000
+        return int(round(random.random() * 2147483647) * cur_ms % 1E10)
+
+    def download_init_data(self, url, mid, fatal=True):
+        webpage = self._download_webpage(url, mid, fatal=fatal)
+        return self._search_json(r'window\.__INITIAL_DATA__\s*=', webpage,
+                                 'init data', mid, transform_source=js_to_json, fatal=fatal)
+
+    def make_fcu_req(self, req_dict, mid, headers={}, **kwargs):
```

Comment:
can we make these private methods?
```suggestion
    def _m_r_get_ruin():
        cur_ms = int(time.time() * 1000) % 1000
        return int(round(random.random() * 2147483647) * cur_ms % 1E10)

    def _download_init_data(self, url, mid, fatal=True):
        webpage = self._download_webpage(url, mid, fatal=fatal)
        return self._search_json(r'window\.__INITIAL_DATA__\s*=', webpage,
                                 'init data', mid, transform_source=js_to_json, fatal=fatal)

    def _make_fcu_req(self, req_dict, mid, headers={}, **kwargs):
```

Reply:
https://github.com/yt-dlp/yt-dlp/pull/9768/commits/393cb6347565a08c0307b41bd4c73618cc17fc44